File: apps/web/lib/objective.ts
/** Display-only projection of gated live publications. Never a research input. */
export const OBJECTIVE_PAGE_SIZE = 20;
export const OBJECTIVE_POLL_MS = 30_000;
export const OBJECTIVE_ORDER = [
  "Cost-adjusted risk/reward descending", "Modelled downside GBP ascending",
  "Cost-adjusted hypothetical upside GBP descending", "Ticker and research ID ascending for reproducible ties",
] as const;

export type Opportunity = {
  research_id: string; ticker: string; company: string; packet_hash: string; snapshot_hash: string; signal_design_hash: string;
  state: "RESEARCH_CANDIDATE"; raw_price: number; raw_currency: "GBP" | "GBX"; normalized_price_gbp: number; conversion_method: string;
  assumed_capital_gbp: number; illustrative_allocation_gbp: number; percentage_of_assumed_capital: number;
  modelled_downside_gbp: number; potential_upside_gbp: number; scenario_return_fraction: number; risk_reward: number;
  round_trip_cost_gbp: number; horizon_days: number; issued_at: string; valid_until: string;
  evidence_independence: "MEDIUM" | "HIGH"; qlib_rank: number | null; qlib_universe_size: number | null; spread_bps: number;
  gap_to_stretch_profit_gbp: number; calibration: "UNCALIBRATED"; probability: null; expected_payoff_gbp: null; evidence_ids: string[];
};
export type ObjectiveRanking = {
  objective: { starting_capital: 200; target_profit: 1000; target_end_value: 1200; target_return: 5; horizon_days: 30; can_override_risk: false };
  basis: "RESEARCH_SIMULATION_NOT_ACTUAL_PORTFOLIO"; coverage: "PAGINATED_WORKSPACE_PUBLICATIONS"; order: readonly string[];
  version: "qualified-scenarios-v1"; evaluated_at: string; examined: number; offset: number; has_more: boolean;
  opportunities: Opportunity[]; limitations: string[]; hash: string;
};

function record(value: unknown): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)) throw new Error("Invalid objective record");
  return value as Record<string, unknown>;
}
function text(value: unknown, maximum = 300): string {
  if (typeof value !== "string" || !value.length || value.length > maximum || /[\u0000-\u001f\u007f]/u.test(value)) throw new Error("Invalid objective text");
  return value;
}
function number(value: unknown, minimum = 0, maximum = 1e12): number {
  if (typeof value !== "number" && (typeof value !== "string" || !/^-?\d+(?:\.\d+)?(?:[Ee][+-]?\d+)?$/.test(value) || value.length > 80)) throw new Error("Invalid objective number");
  const parsed = Number(value);
  if (!Number.isFinite(parsed) || parsed < minimum || parsed > maximum) throw new Error("Objective number outside bounds");
  return parsed;
}
function integer(value: unknown, maximum = 10000): number {
  const parsed = number(value, 0, maximum);
  if (!Number.isInteger(parsed)) throw new Error("Invalid objective integer");
  return parsed;
}
function timestamp(value: unknown): string {
  const parsed = text(value, 40);
  if (!/(?:Z|[+-]\d{2}:\d{2})$/.test(parsed) || !Number.isFinite(Date.parse(parsed))) throw new Error("Invalid objective timestamp");
  return parsed;
}
function hash(value: unknown): string {
  const parsed = text(value, 64);
  if (!/^[a-f0-9]{64}$/.test(parsed)) throw new Error("Invalid objective hash");
  return parsed;
}
function strings(value: unknown, maximum: number, width: number): string[] {
  if (!Array.isArray(value) || value.length > maximum) throw new Error("Invalid objective list");
  return value.map(item => text(item, width));
}

function opportunity(value: unknown): Opportunity {
  const row = record(value);
  const id = text(row.research_id, 36), ticker = text(row.ticker, 32);
  if (!/^[a-f0-9]{8}(?:-[a-f0-9]{4}){3}-[a-f0-9]{12}$/i.test(id) || !/^[A-Z0-9][A-Z0-9._-]{0,31}$/.test(ticker) || ticker === "DEMO.L") throw new Error("Invalid objective identity");
  if (row.state !== "RESEARCH_CANDIDATE" || !["GBP", "GBX"].includes(String(row.raw_currency)) || row.calibration !== "UNCALIBRATED" || row.probability !== null || row.expected_payoff_gbp !== null || !["MEDIUM", "HIGH"].includes(String(row.evidence_independence))) throw new Error("Unqualified objective scenario");
  const capital = number(row.assumed_capital_gbp, Number.MIN_VALUE, 200), allocation = number(row.illustrative_allocation_gbp, Number.MIN_VALUE, capital);
  const raw = number(row.raw_price, Number.MIN_VALUE), normalized = number(row.normalized_price_gbp, Number.MIN_VALUE);
  if (Math.abs(normalized - raw / (row.raw_currency === "GBX" ? 100 : 1)) > Math.max(1e-9, normalized * 1e-10)) throw new Error("Objective currency mismatch");
  const issued = timestamp(row.issued_at), expiry = timestamp(row.valid_until);
  if (Date.parse(expiry) <= Date.parse(issued)) throw new Error("Invalid objective expiry");
  const horizon = integer(row.horizon_days, 30);
  if (horizon < 1) throw new Error("Invalid objective horizon");
  const rank = row.qlib_rank === null ? null : integer(row.qlib_rank, 1e7), universe = row.qlib_universe_size === null ? null : integer(row.qlib_universe_size, 1e7);
  if ((rank === null) !== (universe === null) || (rank !== null && (rank < 1 || universe === null || rank > universe))) throw new Error("Invalid quant rank");
  return {
    research_id: id, ticker, company: text(row.company), packet_hash: hash(row.packet_hash), snapshot_hash: hash(row.snapshot_hash), signal_design_hash: hash(row.signal_design_hash),
    state: "RESEARCH_CANDIDATE", raw_price: raw, raw_currency: row.raw_currency as "GBP" | "GBX", normalized_price_gbp: normalized, conversion_method: text(row.conversion_method),
    assumed_capital_gbp: capital, illustrative_allocation_gbp: allocation, percentage_of_assumed_capital: number(row.percentage_of_assumed_capital, Number.MIN_VALUE, 100),
    modelled_downside_gbp: number(row.modelled_downside_gbp, 0, 200), potential_upside_gbp: number(row.potential_upside_gbp, Number.MIN_VALUE), scenario_return_fraction: number(row.scenario_return_fraction), risk_reward: number(row.risk_reward, Number.MIN_VALUE),
    round_trip_cost_gbp: number(row.round_trip_cost_gbp), horizon_days: horizon, issued_at: issued, valid_until: expiry,
    evidence_independence: row.evidence_independence as "MEDIUM" | "HIGH", qlib_rank: rank, qlib_universe_size: universe, spread_bps: number(row.spread_bps, 0, 10000),
    gap_to_stretch_profit_gbp: number(row.gap_to_stretch_profit_gbp, 0, 1000), calibration: "UNCALIBRATED", probability: null, expected_payoff_gbp: null, evidence_ids: strings(row.evidence_ids, 4000, 256),
  };
}

export function parseObjectiveRanking(value: unknown): ObjectiveRanking {
  const report = record(value), objective = record(report.objective);
  if (report.version !== "qualified-scenarios-v1" || report.basis !== "RESEARCH_SIMULATION_NOT_ACTUAL_PORTFOLIO" || report.coverage !== "PAGINATED_WORKSPACE_PUBLICATIONS" || typeof report.has_more !== "boolean") throw new Error("Invalid objective basis");
  if (number(objective.starting_capital) !== 200 || number(objective.target_profit) !== 1000 || number(objective.target_end_value) !== 1200 || number(objective.target_return) !== 5 || objective.horizon_days !== 30 || objective.can_override_risk !== false) throw new Error("Objective cannot override research boundaries");
  if (!Array.isArray(report.opportunities) || report.opportunities.length > OBJECTIVE_PAGE_SIZE || JSON.stringify(report.order) !== JSON.stringify(OBJECTIVE_ORDER)) throw new Error("Invalid objective ordering");
  const rows = report.opportunities.map(opportunity), examined = integer(report.examined, OBJECTIVE_PAGE_SIZE), evaluated = timestamp(report.evaluated_at);
  if (rows.length > examined || new Set(rows.map(row => row.research_id)).size !== rows.length || rows.some(row => Date.parse(row.issued_at) > Date.parse(evaluated))) throw new Error("Invalid objective coverage");
  return { objective: { starting_capital: 200, target_profit: 1000, target_end_value: 1200, target_return: 5, horizon_days: 30, can_override_risk: false }, basis: "RESEARCH_SIMULATION_NOT_ACTUAL_PORTFOLIO", coverage: "PAGINATED_WORKSPACE_PUBLICATIONS", order: OBJECTIVE_ORDER, version: "qualified-scenarios-v1", evaluated_at: evaluated, examined, offset: integer(report.offset), has_more: report.has_more, opportunities: rows, limitations: strings(report.limitations, 20, 1000), hash: hash(report.hash) };
}

export function objectiveQuery(parameters: URLSearchParams): URLSearchParams | null {
  if ([...parameters.keys()].some(key => !["limit", "offset"].includes(key)) || parameters.getAll("limit").length > 1 || parameters.getAll("offset").length > 1) return null;
  const limit = parameters.get("limit") ?? "20", offset = parameters.get("offset") ?? "0";
  if (!/^[1-9]\d?$/.test(limit) || Number(limit) > 20 || !/^(0|[1-9]\d{0,4})$/.test(offset) || Number(offset) > 10000) return null;
  return new URLSearchParams({ limit, offset });
}

export function currentOpportunities(report: ObjectiveRanking, now: number): Opportunity[] {
  return report.opportunities.filter(row => Date.parse(row.issued_at) <= now && Date.parse(row.valid_until) > now);
}

export type ObjectiveLoad = { report: ObjectiveRanking | null; error: string | null; loading: boolean };
/** Non-overlapping polling; abort and ignore responses from a previous page/session. */
export function observeObjective(offset: number, load: (path: string, signal: AbortSignal) => Promise<unknown>, receive: (state: ObjectiveLoad) => void): () => void {
  const controller = new AbortController();
  let timer: ReturnType<typeof setTimeout> | undefined;
  receive({ report: null, error: null, loading: true });
  async function poll() {
    try {
      const report = parseObjectiveRanking(await load(`/api/research/objective?limit=20&offset=${offset}`, controller.signal));
      if (report.offset !== offset) throw new Error("The research service returned a different page. Please retry.");
      if (!controller.signal.aborted) receive({ report, error: null, loading: false });
    } catch (error) {
      if (!controller.signal.aborted) receive({ report: null, error: error instanceof Error ? error.message : "Objective research is unavailable. Please retry.", loading: false });
    } finally {
      if (!controller.signal.aborted) timer = setTimeout(() => void poll(), OBJECTIVE_POLL_MS);
    }
  }
  void poll();
  return () => { controller.abort(); if (timer) clearTimeout(timer); };
}

File: src/money/research/objective.py
"""Read-only comparisons AFTER publication gates; never an input to research policy.

Ranking is an explicit lexicographic ordering of cost-adjusted scenarios, not a
forecast or an additive AI score. No target, risk limit or allocation is changed
to approach the aspirational objective.
"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import AwareDatetime, Field

from money.policy.governance import consensus, evidence_independence
from money.schemas.contracts import (
    Contract,
    DecisionPacket,
    ResearchState,
    content_hash,
)
from money.signals.generation import SignalDesign, generate_signal


class StretchObjective(Contract):
    """The product objective, with arithmetic fixed independently of candidate selection."""

    starting_capital: Decimal = Field(default=Decimal("200"), ge=200, le=200)
    target_profit: Decimal = Field(default=Decimal("1000"), ge=1000, le=1000)
    target_end_value: Decimal = Field(default=Decimal("1200"), ge=1200, le=1200)
    target_return: Decimal = Field(default=Decimal("5.0"), ge=5, le=5)
    horizon_days: Literal[30] = 30
    can_override_risk: Literal[False] = False


class ObjectivePageQuery(Contract):
    limit: int = Field(default=20, ge=1, le=50)
    offset: int = Field(default=0, ge=0, le=10000)


class OpportunityAssessment(Contract):
    research_id: str
    ticker: str
    company: str
    packet_hash: str
    snapshot_hash: str
    signal_design_hash: str
    state: Literal["RESEARCH_CANDIDATE"] = "RESEARCH_CANDIDATE"
    raw_price: Decimal
    raw_currency: Literal["GBP", "GBX"]
    normalized_price_gbp: Decimal
    conversion_method: str
    assumed_capital_gbp: Decimal = Field(gt=0, le=200)
    illustrative_allocation_gbp: Decimal = Field(gt=0, le=200)
    percentage_of_assumed_capital: Decimal = Field(gt=0, le=100)
    modelled_downside_gbp: Decimal = Field(ge=0, le=200)
    potential_upside_gbp: Decimal = Field(gt=0)
    scenario_return_fraction: Decimal
    risk_reward: Decimal = Field(gt=0)
    round_trip_cost_gbp: Decimal = Field(ge=0)
    horizon_days: int = Field(ge=1, le=30)
    issued_at: AwareDatetime
    valid_until: AwareDatetime
    evidence_independence: Literal["MEDIUM", "HIGH"]
    qlib_rank: int | None
    qlib_universe_size: int | None
    spread_bps: Decimal
    gap_to_stretch_profit_gbp: Decimal = Field(ge=0)
    calibration: Literal["UNCALIBRATED"] = "UNCALIBRATED"
    probability: None = None
    expected_payoff_gbp: None = None
    evidence_ids: tuple[str, ...]


class AsymmetryAnalyzer:
    """Recheck a persisted scenario; high volatility alone cannot qualify it."""

    @staticmethod
    def assess(
        packet: DecisionPacket,
        design: SignalDesign,
        now: datetime,
        *,
        invalidated: bool = False,
    ) -> OpportunityAssessment | None:
        # Revalidation rejects mutated Pydantic instances and modified packet hashes.
        packet = DecisionPacket.model_validate_json(packet.model_dump_json())
        design = SignalDesign.model_validate_json(design.model_dump_json())
        if now.tzinfo is None or now.utcoffset() is None:
            raise ValueError("objective evaluation time must be timezone-aware")
        signal, snapshot = packet.signal, packet.frozen_snapshot
        if (
            invalidated
            or packet.runtime != "live"
            or packet.final_state != ResearchState.RESEARCH_CANDIDATE
            or signal is None
            or snapshot is None
            or packet.issued_at > now
            or signal.issued_at > now
            or signal.effective_state(now) != ResearchState.RESEARCH_CANDIDATE
            or design.signal != signal
            or design.designed_at is None
            or design.designed_at > packet.issued_at
            or design.policy is None
            or design.market_quality is None
            or design.cost_applicability is None
        ):
            return None
        independence = evidence_independence(snapshot, packet.reports)
        state, _ = consensus(
            packet.mandate, snapshot, packet.reports, packet.lean, packet.audit,
            packet.red_team, independence, now, rounds=packet.cross_examination_rounds,
        )
        if state != ResearchState.RESEARCH_CANDIDATE or independence != packet.independence:
            return None
        # The persisted publication path already validates correspondence artifacts.
        # Reproduce the display scenario too: a caller cannot invent upside or costs.
        reproduced = generate_signal(
            packet.research_id, packet.mandate, snapshot, packet.reports, packet.lean,
            packet.audit, packet.red_team, state=state, issued_at=design.designed_at,
            market_quality=design.market_quality, cost_applicability=design.cost_applicability,
            policy=design.policy, rounds=packet.cross_examination_rounds,
        )
        if reproduced != design or not design.targets_gbp or design.costs is None:
            return None
        if (
            design.current_raw is None or design.current_currency is None
            or design.current_gbp is None or design.entry_high_gbp is None
            or design.risk_reward is None or design.market_quality.spread_bps is None
        ):
            return None
        allocation = signal.illustrative_allocation_gbp
        # Use the conservative entry bound and first validated target, not the
        # biggest unvalidated level. No £200 extrapolation of a smaller allocation.
        upside = (
            allocation * (design.targets_gbp[0] / design.entry_high_gbp - 1)
            - design.costs.total_round_trip_gbp
        )
        quant = next(report for report in packet.reports if report.firm == "qlib")
        return OpportunityAssessment(
            research_id=packet.research_id, ticker=signal.ticker,
            company=packet.eligibility.company, packet_hash=packet.hash,
            snapshot_hash=packet.snapshot_hash, signal_design_hash=content_hash(design),
            raw_price=design.current_raw, raw_currency=design.current_currency,
            normalized_price_gbp=design.current_gbp,
            conversion_method=design.normalization_method,
            assumed_capital_gbp=signal.assumed_capital_gbp,
            illustrative_allocation_gbp=allocation,
            percentage_of_assumed_capital=allocation / signal.assumed_capital_gbp * 100,
            modelled_downside_gbp=signal.modelled_downside_gbp,
            potential_upside_gbp=upside, scenario_return_fraction=upside / allocation,
            risk_reward=design.risk_reward,
            round_trip_cost_gbp=design.costs.total_round_trip_gbp,
            horizon_days=signal.horizon_days, issued_at=signal.issued_at,
            valid_until=signal.valid_until,
            evidence_independence="HIGH" if independence.strength == "HIGH" else "MEDIUM",
            qlib_rank=getattr(quant, "rank", None),
            qlib_universe_size=getattr(quant, "universe_size", None),
            spread_bps=design.market_quality.spread_bps,
            gap_to_stretch_profit_gbp=max(Decimal(0), StretchObjective().target_profit - upside),
            evidence_ids=design.evidence_ids,
        )


class OpportunityRanking(Contract):
    version: Literal["qualified-scenarios-v1"] = "qualified-scenarios-v1"
    objective: StretchObjective = Field(default_factory=StretchObjective)
    evaluated_at: AwareDatetime
    basis: Literal["RESEARCH_SIMULATION_NOT_ACTUAL_PORTFOLIO"] = (
        "RESEARCH_SIMULATION_NOT_ACTUAL_PORTFOLIO"
    )
    coverage: Literal["PAGINATED_WORKSPACE_PUBLICATIONS"] = "PAGINATED_WORKSPACE_PUBLICATIONS"
    examined: int = Field(ge=0)
    offset: int = Field(ge=0)
    has_more: bool
    opportunities: tuple[OpportunityAssessment, ...]
    conclusion: Literal[
        "NO_QUALIFIED_OPPORTUNITY_CURRENTLY_SUPPORTS_THE_STRETCH_OBJECTIVE",
        "MODELLED_SCENARIO_REACHES_OBJECTIVE_WITHOUT_CALIBRATED_PROBABILITY",
    ]
    order: tuple[str, ...] = (
        "Cost-adjusted risk/reward descending",
        "Modelled downside GBP ascending",
        "Cost-adjusted hypothetical upside GBP descending",
        "Ticker and research ID ascending for reproducible ties",
    )
    limitations: tuple[str, ...] = (
        "Ranked only within this page of this workspace's sealed live publications, not the entire market.",
        "Targets are hypothetical scenarios, not expected returns or calibrated probabilities.",
        "No qualified scenario is evidence that a +500% return will occur.",
        "Alternatives are not a portfolio: allocations must not be added together.",
        "No actual trades, portfolio value, countdown or reinvested compounding is inferred.",
        "Gaps can exceed modelled downside. Investment decisions remain human.",
    )
    hash: str = ""


def rank_opportunities(
    assessments: tuple[OpportunityAssessment, ...], *, now: datetime,
    examined: int, offset: int = 0, has_more: bool = False,
) -> OpportunityRanking:
    """Compare already gated alternatives; the stretch objective never affects ordering."""
    ordered = tuple(sorted(assessments, key=lambda item: (
        -item.risk_reward, item.modelled_downside_gbp, -item.potential_upside_gbp,
        item.ticker, item.research_id,
    )))
    report = OpportunityRanking(
        evaluated_at=now, examined=examined, offset=offset, has_more=has_more,
        opportunities=ordered,
        conclusion=(
            "MODELLED_SCENARIO_REACHES_OBJECTIVE_WITHOUT_CALIBRATED_PROBABILITY"
            if any(item.gap_to_stretch_profit_gbp == 0 for item in ordered)
            else "NO_QUALIFIED_OPPORTUNITY_CURRENTLY_SUPPORTS_THE_STRETCH_OBJECTIVE"
        ),
    )
    return report.model_copy(update={
        "hash": content_hash(report.model_dump(mode="json", exclude={"hash"})),
    })
